X86/X86MachineInstructionRaiser.cpp

```diff
@@ -2149,6 +2140,20 @@ bool X86MachineInstructionRaiser::raiseMoveToMemInstr(const MachineInstr &MI,
       StoreTy = Type::getDoubleTy(Ctx);
   }
 
+  LoadInst *LdInst = nullptr;
+  if (!isMovInst) {
+    if (MemRefVal->getType()->getPointerElementType() != StoreTy) {
+      MemRefVal = new BitCastInst(
+          MemRefVal, StoreTy->getPointerTo(), "", RaisedBB);
+    }
+
+    // Load the value from memory location
+    auto align =
+        MemRefVal->getPointerAlignment(MR->getModule()->getDataLayout());
+    LdInst = new LoadInst(MemRefVal->getType()->getPointerElementType(),
+                          MemRefVal, "", false, align, RaisedBB);
+  }
+
   // Cast SrcValue and MemRefVal as needed.
   MemRefVal = getRaisedValues()->castValue(MemRefVal, StoreTy->getPointerTo(),
```

Comment:
Thanks @martin-fink for adding the test.

`MemRefVal`is being already cast here, as appropriate. So, the following change - instead of the one in this commit - addresses the failure exposed by the test added.

```
diff --git a/X86/X86MachineInstructionRaiser.cpp b/X86/X86MachineInstructionRaiser.cpp
index 142d431..b55fa36 100644
--- a/X86/X86MachineInstructionRaiser.cpp
+++ b/X86/X86MachineInstructionRaiser.cpp
@@ -2127,15 +2127,6 @@ bool X86MachineInstructionRaiser::raiseMoveToMemInstr(const MachineInstr &MI,
     MemRefVal = convIntToPtr;
   }
 
-  LoadInst *LdInst = nullptr;
-  if (!isMovInst) {
-    // Load the value from memory location
-    auto align =
-        MemRefVal->getPointerAlignment(MR->getModule()->getDataLayout());
-    LdInst = new LoadInst(MemRefVal->getType()->getPointerElementType(),
-                          MemRefVal, "", false, align, RaisedBB);
-  }
-
   // This instruction moves a source value to memory. So, if the types of
   // the source value and that of the memory pointer element are not the
   // same as that of the store size of the instruction, cast them as needed.
@@ -2157,6 +2148,12 @@ bool X86MachineInstructionRaiser::raiseMoveToMemInstr(const MachineInstr &MI,
   SrcValue = getRaisedValues()->castValue(SrcValue, StoreTy, RaisedBB);
 
   if (!isMovInst) {
+    // Load the value from memory location
+    auto align =
+        MemRefVal->getPointerAlignment(MR->getModule()->getDataLayout());
+    LoadInst *LdInst =
+        new LoadInst(MemRefVal->getType()->getPointerElementType(), MemRefVal,
+                     "", false, align, RaisedBB);
     // If this is not an instruction that just moves SrcValue, generate the
     // instruction that performs the appropriate operation and then store the
     // result in MemRefVal.
```

Let me know of any feedback you might have.

Reply:
I've applied this change instead of mine, thanks!